lib/active_merchant/billing/gateways/ipg.rb

```diff
@@ -0,0 +1,363 @@
+        xml.tag!('v1:recurringType', recurring_type)
+      end
+
+      def add_storage_item(xml, credit_card, options)
+        requires!(options.merge!({ credit_card: credit_card }), :credit_card, :hosted_data_id)
+        xml.tag!('ns2:StoreHostedData') do
+          xml.tag!('ns2:DataStorageItem') do
+            add_payment(xml, credit_card, {}, 'ns2')
+            add_three_d_secure(xml, options[:three_d_secure]) if options[:three_d_secure]
+            xml.tag!('ns2:HostedDataID', options[:hosted_data_id]) if options[:hosted_data_id]
+          end
+        end
+      end
+
+      def add_transaction_type(xml, type)
+        xml.tag!('v1:CreditCardTxType') do
+          xml.tag!('v1:StoreId', @credentials[:store_id])
+          xml.tag!('v1:Type', type)
+        end
+      end
+
+      def add_payment(xml, payment, options = {}, credit_envelope = 'v1')
+        requires!(options.merge!({ card_number: payment.number, month: payment.month, year: payment.year }), :card_number, :month, :year) if payment
+        if payment
+          xml.tag!("#{credit_envelope}:CreditCardData") do
+            xml.tag!('v1:CardNumber', payment.number) if payment.number
+            xml.tag!('v1:ExpMonth', payment.month) if payment.month
+            xml.tag!('v1:ExpYear', payment.year) if payment.year
+            xml.tag!('v1:CardCodeValue', payment.verification_value) if payment.verification_value
+            xml.tag!('v1:Brand', options[:brand]) if options[:brand]
+          end
+        end
+
+        if options[:card_function_type]
+          xml.tag!('v1:cardFunction') do
+            xml.tag!('v1:Type', options[:card_function_type])
+          end
+        end
+
+        if options[:track_data]
+          xml.tag!("#{credit_envelope}:CreditCardData") do
+            xml.tag!('v1:TrackData', options[:track_data])
+          end
+        end
+      end
+
+      def add_three_d_secure(xml, three_d_secure)
+        xml.tag!('v1:CreditCard3DSecure') do
+          xml.tag!('v1:AuthenticationValue', three_d_secure[:cavv]) if three_d_secure[:cavv]
+          xml.tag!('v1:XID', three_d_secure[:xid]) if three_d_secure[:xid]
+          xml.tag!('v1:Secure3D2TransactionStatus', three_d_secure[:directory_response_status]) if three_d_secure[:directory_response_status]
+          xml.tag!('v1:Secure3D2AuthenticationResponse', three_d_secure[:authentication_response_status]) if three_d_secure[:authentication_response_status]
+          xml.tag!('v1:Secure3DProtocolVersion', three_d_secure[:version]) if three_d_secure[:version]
+          xml.tag!('v1:DirectoryServerTransactionId', three_d_secure[:ds_transaction_id]) if three_d_secure[:ds_transaction_id]
+        end
+      end
+
+      def add_transaction_details(xml, options, pre_order = false)
+        requires!(options, :OrderId) if pre_order
+        xml.tag!('v1:TransactionDetails') do
+          xml.tag!('v1:OrderId', options[:OrderId]) if options[:OrderId]
```

Comment:
We generally expect options to be passed in snake_case rather than CamelCase; can we do that here?